js/cards/langcard.ts

```diff
@@ -0,0 +1,225 @@
+/* Copyright 2010-2019 Will Scullin <scullin@scullinsteel.com>
+ *
+ * Permission to use, copy, modify, distribute, and sell this software and its
+ * documentation for any purpose is hereby granted without fee, provided that
+ * the above copyright notice appear in all copies and that both that
+ * copyright notice and this permission notice appear in supporting
+ * documentation.  No representations are made about the suitability of this
+ * software for any purpose.  It is provided "as is" without express or
+ * implied warranty.
+ */
+
+import RAM, { RAMState } from '../ram';
+import { debug } from '../util';
+import { Card, Memory, byte, Restorable } from '../types';
+
+export interface LanguageCardState {
+    bank1: RAMState;
+    bank2: RAMState;
+    ram: RAMState;
+
+    readbsr: boolean;
+    writebsr: boolean;
+    bsr2: boolean;
+    prewrite: boolean;
+}
+
+export default class LanguageCard implements Card, Restorable<LanguageCardState> {
+    LOC = {
+        // Bank 2
+        READBSR2: 0x80,
+        WRITEBSR2: 0x81,
+        OFFBSR2: 0x82,
+        READWRBSR2: 0x83,
+
+        // Bank 1
+        READBSR1: 0x88,
+        WRITEBSR1: 0x89,
+        OFFBSR1: 0x8a,
+        READWRBSR1: 0x8b,
+    }
+
+    private bank1: RAM;
+    private bank2: RAM;
+    private ram: RAM;
+
+    private readbsr = false;
+    private writebsr = false;
+    private bsr2 = false;
+    private prewrite = false;
+
+    private read1: Memory;
+    private read2: Memory;
+
+    private write1: Memory;
+    private write2: Memory;
+
+    constructor(private rom: Memory) {
+        debug('Language card');
+
+        this.bank1 = new RAM(0xd0, 0xdf);
+        this.bank2 = new RAM(0xd0, 0xdf);
+        this.ram = new RAM(0xe0, 0xff);
+
+        this.write1 = this.rom;
+        this.write2 = this.rom;
+
+        this.read1 = this.rom;
+        this.read2 = this.rom;
+    }
+
+    private debug(..._args: any[]) {
+        // debug.apply(null, arguments);
+    }
+
+
+    updateBanks() {
+        if (this.readbsr) {
+            this.read1 = this.bsr2 ? this.bank2 : this.bank1;
+            this.read2 = this.ram;
+        } else {
+            this.read1 = this.rom;
+            this.read2 = this.rom;
+        }
+
+        if (this.writebsr) {
+            this.write1 = this.bsr2 ? this.bank2 : this.bank1;
+            this.write2 = this.ram;
+        } else {
+            this.write1 = this.rom;
+            this.write2 = this.rom;
+        }
+    }
+
+    private access(off: byte, val: byte) {
+        const readMode = val === undefined;
+        const result = readMode ? 0 : undefined;
+
+        switch (off & 0x8B) {
```

Comment:
I've always known it could be boiled down to a handful of conditionals, but I've always struggled with getting it just right and the case statement just worked. But I've given it another go and it seems to work now.